relay-server/src/actors/project.rs

```diff
@@ -421,6 +424,25 @@ impl Project {
         }
     }
 
+    pub fn reservoir_counters(&self) -> ReservoirCounters {
+        self.reservoir_counters.clone()
+    }
+
+    /// if a reservoir rule is no longer in the sampling config, we can assume it's no longer needed.
+    pub fn remove_expired_reservoir_rules(&self) {
+        let Some(config) = self
+            .state
+            .as_ref()
+            .and_then(|state| state.config.dynamic_sampling.as_ref())
+        else {
+            return;
+        };
+
+        if let Ok(mut guard) = self.reservoir_counters.try_lock() {
+            guard.retain(|key, _| config.rules_v2.iter().any(|rule| rule.id == *key));
```

Comment:
@jan-auer  avoiding allocations by simply passing in all the rules, regardless of if they're reservoir or not. Shouldn't matter at all since non-reservoir ones wont match anyway.